File: src/page.js
const puppeteer = require('puppeteer');

async function init() {
  try {
    const browser = await puppeteer.launch({ headless: true });
    console.log('Browser is ready...');
    return browser;
  } catch (e) {
    console.log('Failed to start browser.');
    process.exit(1);
  }
}

async function open(browser, url) {
  let page;
  try {
    page = await browser.newPage();
    await page.goto(url);

    return await page.$$eval('g-scrolling-carousel a[aria-label]', (carousels) => {
      return carousels.map(carousel => {
        const extractedInfo = {
          "name": carousel.ariaLabel,
          "link": `${carousel.href.replace('file://', 'https://www.google.com')}`,
        };
        const extensionText = carousel.innerText
          .replace('-\n', '-') // title in the card breaks at '-', eg, line1: Service-\n, line2: Orient...
          .replace('\n', ' ')
          .replace(carousel.ariaLabel, '').trim();
        if(extensionText) extractedInfo.extensions = [ extensionText ];
        const imgElement = carousel.querySelector('img');
        if(imgElement.dataset.lzy_) { 
          // lazy loaded image, set to null
          extractedInfo.image = null;
        } else {
          extractedInfo.image = imgElement.src || null;
        }
        return extractedInfo;
      });
    });
  } catch(e) {
    console.log(e);
    throw e;
  } finally {
    if(page != null) {
      await page.close();
    }
  }
}
module.exports = { init, open };

File: page.rb
#!/usr/bin/ruby

require "nokogiri"
require "json"

class Page
  def initialize(file)
    @artworks = []
    @parsed_page ||= File.open(file) { |f| Nokogiri::HTML(f) }
  end
    
  def parse()
    @parsed_page.css("g-scrolling-carousel a[aria-label]").each do |carousel|
      name = carousel["aria-label"]
      link = carousel["href"].start_with?("https://www.google.com") ? carousel["href"] : "https://www.google.com" + carousel["href"]
      extensionText = carousel.text
        .gsub("-\n", "-") # title in the card breaks at "-", eg, line1: Service-\n, line2: Orient...
        .gsub("\n", " ")
        .gsub(name, "")
      extensionText.strip!

      parsed = {
        name: name,
        link: link,
      }

      if(extensionText.length > 0)
        parsed[:extensions] = [ extensionText ]
      end

      img_node = carousel.css("img")[0]
      # TODO: not able to get img_node["data-lzy_"]
      parsed[:image] = img_node["src"]

      @artworks << parsed
    end   

    JSON.generate(@artworks)
  end  
end
